### PR TITLE
Feature/support hsm

### LINKED ISSUE #4992: Support for Hardware Secure Module (HSM) integration in crypto layer
#### Problem
- No support for Hardware Secure Module (HSM) integration in crypto layer
- Allow required Crypto operations to be offloaded to HSM for additional security/performance
- Allow secure storage of credentials in HSM
- Option to switch between HSM and Host Software Crypto implementation required
 
#### Proposed Solution
- Definition and Implementation of HSM API in src/crypto/hsm
- HSM API called from the existing crypto modules
- config to switch between HSM & Host Software Crypto implementation

### PR DESCRIPTION
**Pull request by Jagadish-NXP**
#### Problem
 Support for Hardware Secure Module (HSM) integration in crypto layer

 #### Summary of Changes
- Hardware Secure Module (HSM) integration in crypto layer
- Defined and Implemented HSM API in src/crypto/hsm
- Allow required Crypto operations to be offloaded to HSM for additional security/performance
- HSM API are called from the existing crypto modules
- config to switch between HSM & Host Software Crypto implementation

 Fixes #4992

### CODE REVIEW

File: src/crypto/CHIPCryptoPAL.cpp [hunk 1]
```diff
@@ -21,6 +21,7 @@
  */
 
 #include "CHIPCryptoPAL.h"
+#include "hsm/CHIPCryptoPALHsm.h"
```
Comment (mdamle) [inline src/crypto/CHIPCryptoPAL.cpp:24]: Protect the header include with #ifdef CHIP_CRYPTO_HSM

File: src/crypto/CHIPCryptoPAL.h [hunk 2]
```diff
@@ -30,6 +30,8 @@
 #include <core/CHIPError.h>
 #include <support/CodeUtils.h>
 
+#include "hsm/CHIPCryptoPALHsm_config.h"
```
Comment (mdamle) [inline src/crypto/CHIPCryptoPAL.h:33]: Same as above. Consider protecting header include with #if CHIP_CRYPTO_HSM == 1

File: src/crypto/CHIPCryptoPAL.cpp (hunk 1 above)
Comment (Jagadish-NXP) [reply src/crypto/CHIPCryptoPAL.cpp:26]: Sure. Now protected

File: src/crypto/CHIPCryptoPAL.h (hunk 2 above)
Comment (Jagadish-NXP) [reply src/crypto/CHIPCryptoPAL.h:34]: Sure This Inclusion is also protected

File: src/crypto/hsm/nxp/README.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+## Introduction
+
+CHIPCryptoPALHsm_SE05X.cpp file provides the integration of secure element (SE05x) in crypto layer of CHIP stack.
+By enabling the required directives in CHIPCryptoPALHsm_config.h header file,  required crypto operations can be performed using SE05x.
+By default, the secure element is enabled only for PAKE (Verifier role) operations.
+
+Following table shows the operations supported by secure element.
+
++-------------------------+-----------+----------------+---------------------------------+------------+----------------------------+
+| Operation               | Software  | Default config | Note                            | Re-entrant | Directive to enabled in    |
```
Comment (msandstedt) [inline src/crypto/hsm/nxp/README.md:12]: I'm glad reentrancy is documented.  However, can we not add locking to the new critical paths to maintain thread safety for `Spake2p` the session objects that use this?

File: src/crypto/crypto.gni [hunk 4]
```diff
@@ -15,6 +16,7 @@
 declare_args() {
   # Crypto implementation: mbedtls, openssl.
   chip_crypto = ""
+  with_se05x = 0
```
Comment (andy31415) [inline src/crypto/crypto.gni:19]: I think we standardise to using `chip_` as a prefix for all gn arguments added by chip.

File: src/crypto/hsm/nxp/README.md (hunk 3 above)
Comment (Jagadish-NXP) [reply src/crypto/hsm/nxp/README.md:12]: Thanks. Let me have a closer look on how best to handle this.

Comment (msandstedt) [reply src/crypto/hsm/nxp/README.md]: I would think the most straight forward thing to do would be to add a static mutex to the `Spake2p` definition.  Then calls to the HSM can either block on the mutex, or a try-lock approach could used to fall back to software implementations.

The former approach enforces use of the HSM always.  The latter approach gives an opportunity for concurrency.

Comment (Jagadish-NXP) [reply src/crypto/hsm/nxp/README.md]: We now have taken care of Re-entrancy for the Spake2+ HSM based implementation. We moved a static data structure into the Spake Class. A Mutex is now used to protect a shared object

Comment (msandstedt) [reply src/crypto/hsm/nxp/README.md]: Thank you!

**Review by msandstedt — APPROVED** [review]

File: .gitmodules [hunk 5]
```diff
@@ -74,3 +74,7 @@
 	path = third_party/efr32_sdk/repo
 	url = https://github.com/SiliconLabs/sdk_support.git
 	branch = master
+[submodule "simw-top-mini"]
```
Comment (andy31415) [inline .gitmodules:77]: the submodule name seems very different than the Github URL. why is that? should we name it 'plug_and_trust' or similar instead?

File: src/crypto/CHIPCryptoPAL.cpp [hunk 6]
```diff
@@ -188,6 +256,33 @@ CHIP_ERROR Spake2p::ComputeRoundOne(uint8_t * out, size_t * out_len)
     VerifyOrExit(state == CHIP_SPAKE2P_STATE::STARTED, error = CHIP_ERROR_INTERNAL);
     VerifyOrExit(*out_len >= point_size, error = CHIP_ERROR_INTERNAL);
 
+#if ((CHIP_CRYPTO_HSM) && (!ENABLE_HSM_SPAKE_VERIFIER))
```
Comment (andy31415) [inline src/crypto/CHIPCryptoPAL.cpp:259]: I find these ifdefs hard to follow. Could we ifdef entire function implementations instead ? Alternatively have some helper `IsMissingImplementation(role)` and narrow ifdefs in there.

However really it seems we have eitherHSM implementation or nonHSM.

File: src/crypto/crypto.gni (hunk 4 above)
Comment (Jagadish-NXP) [reply src/crypto/crypto.gni]: Sure it is now chip_with_se05x

File: .gitmodules (hunk 5 above)
Comment (Jagadish-NXP) [reply .gitmodules:77]: Plug & Trust Middleware is a software package supporting several different microcontrollers on different platforms. The full package is distributed via nxp.com we provide a reduced package (simw-top-mini) under an open-source license to facilitate integration with the CHIP build system.
Preferably, We would like to retain current naming to be consistent between the different packages and its documentation.

File: src/crypto/CHIPCryptoPAL.cpp (hunk 6 above)
Comment (Jagadish-NXP) [reply src/crypto/CHIPCryptoPAL.cpp:259]: Do you have a suggestion here? If we ifdef entire functions, wouldn't it  be too much code duplication. For the second approach, as the ifdefs wraps different codes at different places, IsMissingImplementation(role) would mean 'different' at different places. Is my understanding correct or am I missing it?

Comment (andy31415) [reply src/crypto/CHIPCryptoPAL.cpp:259]: I found it hard to judge how much is duplicated - at a first glance it seemed the code paths were completely separate except some precondition/argument checks. I was hoping that we could have completely separated out logic.

It could be done as a followup if you believe it is impractical. many separate `#ifdef` does make code review difficult.

Please apply `./scripts/helpers/restyle-diff.sh` to resolve restyle errors. This will add more reviewers and the PR will not show up as red in open PRs.

Comment (Jagadish-NXP) [reply src/crypto/CHIPCryptoPAL.cpp:259]: Thanks I will apply ./scripts/helpers/restyle-diff.sh. I was not aware of it. Regarding ifdef, We can separate it for sure as a follow-up PR, But one other thing that complicates the separation is that there is fallback option to software at certain places. But will definitely look into it

**Review by andy31415 — APPROVED** [review]

File: src/crypto/CHIPCryptoPAL.cpp (hunk 1 above)
Comment (Jagadish-NXP) [reply src/crypto/CHIPCryptoPAL.cpp:26]: Please review and let me know if the solution is acceptable

**Comment by Jagadish-NXP** [comment]
Can somebody please help to identify what is causing this failure. How to run this test locally?

**Review by jelderton — APPROVED** [review]

File: src/crypto/CHIPCryptoPAL.cpp [hunk 7]
```diff
@@ -123,9 +133,33 @@ CHIP_ERROR Spake2p::BeginVerifier(const uint8_t * my_identity, size_t my_identit
                                   size_t peer_identity_len, const uint8_t * w0in, size_t w0in_len, const uint8_t * Lin,
                                   size_t Lin_len)
 {
+#if ((CHIP_CRYPTO_HSM) && (ENABLE_HSM_SPAKE_VERIFIER))
+    uint8_t w0in_mod[32] = {
+        0,
+    };
+    size_t w0in_mod_len = 32;
```
Comment (bzbarsky-apple) [inline src/crypto/CHIPCryptoPAL.cpp:140]: We have this in a bunch of places, and it looks like it can cause things to get out of sync, right?  Why not:
```
constexpr size_t w0in_mod_len = 32;
uint8_t w0in_mod[w0in_mod_len] = ....
```
And maybe even add template overloads of FELoad/FEWrite that will auto-deduce the length when a C array is used?  Followup for this, please?  @Jagadish-NXP

File: src/crypto/CHIPCryptoPAL.cpp [hunk 8]
```diff
@@ -123,9 +133,33 @@ CHIP_ERROR Spake2p::BeginVerifier(const uint8_t * my_identity, size_t my_identit
...
+#endif
     CHIP_ERROR error = CHIP_ERROR_INTERNAL;
     VerifyOrExit(state == CHIP_SPAKE2P_STATE::INIT, error = CHIP_ERROR_INTERNAL);
 
+#if ((CHIP_CRYPTO_HSM) && (ENABLE_HSM_SPAKE_VERIFIER))
+
+    error = FELoad(w0in, w0in_len, w0);
```
Comment (bzbarsky-apple) [inline src/crypto/CHIPCryptoPAL.cpp:147]: This could use a comment explaining why we can't just use `w0in`.....

**Review by bzbarsky-apple — COMMENTED** [review]
I have the same general question as Andrei wrt the ifdefs.  For a lot of these it seems like using a subclass that overrides the relevant functions, or a crypto PAL impl that is initialized with a software impl to delegate to, and then decides whether to delegate to the software imple or not would be a lot more readable....  For things like `AES_CCM_encrypt` the second approach (i.e. not subclass) would let us also avoid duplicating logic between openssl and mbedtls for the HSM case....

And to be clear, I don't know whether we can eliminate _all_ ifdefs by structuring things in some better way here, but I am sure we can eliminate quite a number of them.  And then we can see what's left and evaluate....

**Review by andy31415 — DISMISSED** [review]
Asking for changes: please remove extra copyrights. Oddly enough changed files got NXP on them now and new files only keep chip authors. I would go with chip authors everywhere.

Also highlighting concerns regarding maintainability - since I am not the only one I would give higher importance to have the code organized with less #ifdef branches.

**Review by pan-apple — DISMISSED** [review]
This PR is adding #ifdefs to `OpenSSL` and `mbedTLS` CHIP crypto implementations. Similar wrappers will be needed in every abstraction layer that might be added in future.

I think, we should be able to derive/specialize `P256PublicKey` to say `HSMP256PublicKey` and override necessary methods. The overridden functions can call super class functions if HSM calls fail.

#ifdefs can be used at places where `P256PublicKey` is current being instantiated, to instead instantiate `HSMP256PublicKey`. If the code is repeated, maybe a wrapper can be added to hide the #ifdef.

Similar patterns can be done for other Class based data types. Some of the functions (e.g. `AES_CCM_encrypt`, `AES_CCM_decrypt`, and `Hash_SHA256` are C APIs. Those can eventually (may be in a follow on PR) be moved to a CPP style API.

File: src/crypto/BUILD.gn [hunk 9]
```diff
@@ -1,4 +1,5 @@
 # Copyright (c) 2020 Project CHIP Authors
+# Copyright 2020 NXP
```
[the remark this answers is not in the document]
Comment (Jagadish-NXP) [reply src/crypto/BUILD.gn:2]: Sure. This will be pushed along with the other changes

File: src/crypto/crypto.gni [hunk 10]
```diff
@@ -1,4 +1,5 @@
 # Copyright (c) 2020 Project CHIP Authors
+# Copyright 2020 NXP
```
[the remark this answers is not in the document]
Comment (Jagadish-NXP) [reply src/crypto/crypto.gni:2]: Sure. This will be pushed along with the other changes

File: src/crypto/CHIPCryptoPAL.cpp (hunk 7 above)
Comment (Jagadish-NXP) [reply src/crypto/CHIPCryptoPAL.cpp:140]: regarding the ifdefs, point taken. We will be doing it the way pan-apple suggested below

File: src/crypto/BUILD.gn (hunk 9 above)
[the remark this answers is not in the document]
Comment (Jagadish-NXP) [reply src/crypto/BUILD.gn:2]: Agreed

File: src/crypto/crypto.gni (hunk 10 above)
[the remark this answers is not in the document]
Comment (Jagadish-NXP) [reply src/crypto/crypto.gni:2]: Agreed

**Comment by Jagadish-NXP** [comment]
> This PR is adding #ifdefs to `OpenSSL` and `mbedTLS` CHIP crypto implementations. Similar wrappers will be needed in every abstraction layer that might be added in future.
> 
> I think, we should be able to derive/specialize `P256PublicKey` to say `HSMP256PublicKey` and override necessary methods. The overridden functions can call super class functions if HSM calls fail.
> 
> #ifdefs can be used at places where `P256PublicKey` is current being instantiated, to instead instantiate `HSMP256PublicKey`. If the code is repeated, maybe a wrapper can be added to hide the #ifdef.
> 
> Similar patterns can be done for other Class based data types. Some of the functions (e.g. `AES_CCM_encrypt`, `AES_CCM_decrypt`, and `Hash_SHA256` are C APIs. Those can eventually (may be in a follow on PR) be moved to a CPP style API.

Agreed. Will do it as suggested

**Comment by Jagadish-NXP** [comment]
We are working on restructuring the code as per the review comments. Will push the updates shortly

**Comment by Jagadish-NXP** [comment]
As Suggested
We have removed the #ifdefs to OpenSSL and mbedTLS CHIP crypto implementations.
We are deriving the spake class and overriding necessary methods.
We are deriving only the Spake class for HSM in this PR. Other required crypto classes will be introduced in subsequent PRs
Also the extra copyrights have been removed

**Comment by Jagadish-NXP** [comment]
> Asking for changes: please remove extra copyrights. Oddly enough changed files got NXP on them now and new files only keep chip authors. I would go with chip authors everywhere.
> 
> Also highlighting concerns regarding maintainability - since I am not the only one I would give higher importance to have the code organized with less #ifdef branches.

Done as suggested. Please have a look

**Comment by Jagadish-NXP** [comment]
> This PR is adding #ifdefs to `OpenSSL` and `mbedTLS` CHIP crypto implementations. Similar wrappers will be needed in every abstraction layer that might be added in future.
> 
> I think, we should be able to derive/specialize `P256PublicKey` to say `HSMP256PublicKey` and override necessary methods. The overridden functions can call super class functions if HSM calls fail.
> 
> #ifdefs can be used at places where `P256PublicKey` is current being instantiated, to instead instantiate `HSMP256PublicKey`. If the code is repeated, maybe a wrapper can be added to hide the #ifdef.
> 
> Similar patterns can be done for other Class based data types. Some of the functions (e.g. `AES_CCM_encrypt`, `AES_CCM_decrypt`, and `Hash_SHA256` are C APIs. Those can eventually (may be in a follow on PR) be moved to a CPP style API.

Implemented as suggested. Please have a look

**Comment by Jagadish-NXP** [comment]
> I have the same general question as Andrei wrt the ifdefs. For a lot of these it seems like using a subclass that overrides the relevant functions, or a crypto PAL impl that is initialized with a software impl to delegate to, and then decides whether to delegate to the software imple or not would be a lot more readable.... For things like `AES_CCM_encrypt` the second approach (i.e. not subclass) would let us also avoid duplicating logic between openssl and mbedtls for the HSM case....
> 
> And to be clear, I don't know whether we can eliminate _all_ ifdefs by structuring things in some better way here, but I am sure we can eliminate quite a number of them. And then we can see what's left and evaluate....

We have restructured the code and eliminated the ifdefs in the OpenSSL and mbedTLS CHIP crypto implementation. Please have a look

File: src/crypto/hsm/CHIPCryptoPALHsm.h [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+public:
+    Spake2pHSM_P256_SHA256_HKDF_HMAC() {}
+
+    ~Spake2pHSM_P256_SHA256_HKDF_HMAC() {}
+
+    CHIP_ERROR Init(const uint8_t * context, size_t context_len);
+
+#if ENABLE_HSM_SPAKE_VERIFIER
+    CHIP_ERROR BeginVerifier(const uint8_t * my_identity, size_t my_identity_len, const uint8_t * peer_identity,
+                             size_t peer_identity_len, const uint8_t * w0in, size_t w0in_len, const uint8_t * Lin, size_t Lin_len);
```
Comment (pan-apple) [inline src/crypto/hsm/CHIPCryptoPALHsm.h:57]: Wondering why the compiler is not complaining about missing `override` at the end of these functions.

**Review by woody-apple — APPROVED** [review]
